0000-scheduled-circuit.md

```diff
@@ -0,0 +1,266 @@
+as well as scheduling methods to construct it from a QuantumCircuit.
+
+### Main use cases
+1) accurate simulation of decoherence in Aer (T1&T2, Cross-talk)
+2) insert gates in idle timeslots
+
+## User Benefit
+- Time-aware circuit-based noise simulation on Aer is enabled.
+- Scheduling in circuit level will enable more optimization than that in pulse level in the future.
+
+## Design Proposal
+### High-level API:
+1. Noise simulation flow (new):
+```
+timed_circuit = circuit.schedule(transpiled_circuit, ...)  # transpiled_circuit: QuantumCircuit
+qobj = assemble(timed_circuit)  # timed_circuit: ScheduledCircuit
+```
+
+2. Circuit-to-Schedule flow (update):
+```
+schedule = scheduler.circuit_schedule(transpiled_circuit, ...)  # transpiled_circuit: QuantumCircuit
+```
+->
+```
+timed_circuit = circuit.schedule(transpiled_circuit, ...)  # transpiled_circuit: QuantumCircuit
+schedule = pulse.build_schedule(timed_circuit, ...)  # timed_circuit: ScheduledCircuit
+```
+
+### Requirements:
+- It needs to mimic the scheduling that would be run on a real device as closely as possible
+- It needs serialize to standard qasm qobj, but with a `duration` (or `gate_length`) field to each instruction
+   all qubits should have an instruction assigned to them at all times so:
+- It needs to insert `delay`/wait/idle instructions on any qubit that is idle
+   (with the duration saying how long it is idle for)
+- It needs to support some interfaces for inserting gates to idle timeslots.
+- [For future (cross-talk simulation on Aer)] It may need to support optional insertion of all-qubit time-step instructions at each layer,
+   which would also serialize to a qasm qobj instruction with a duration value given by the total time of the time-step layer.
+
+
+## Detailed Design
+Technical reference level design. Elaborate on details such as:
+- Implementation procedure
+  - If spans multiple projects cover these parts individually
+- Interaction with other features
+- Dissecting corner cases
+- Reference definition, eg., formal definitions.
+
+### Example code
+```
+def schedule(qc: QuantumCircuit,
+             backend_props: BackendProperties,
+             method: str = "alap",
+             dependency: Optional[DependencyGraph] = None) -> ScheduledCircuit:
+    """
+    Args:
+        qc: a transpiled circuit to be scheduled.
+        backend_props: to get gate_length
+        method: same as in pulse.scheduler.schedule_circuit
+        dependency: (for future) asap and alap can work without this
+    Returns:
+        The scheduled circuit.
+    """
+    if dependency is None:
+        # schedule based on the standard dependency that is the same as used in DAGCircuit
+        # algorithm is almost the same as in pulse.scheduler.methods.algorithm
+    else:
+        # scheudle based on the extended dependency for the future
+
+
+Qubit = int # assume only physical qubits
+ArgumentedInstruction = Tuple[Instruction, List[Qubit], List[Clbit]]
```

Comment:
```suggestion
ArgumentedInstruction = Tuple[Instruction, List[Qubit], List[Clbit]]
```

Would it be possible for this to become standard for operands to be accepted in the instruction definition? Having a tuple format is really limiting in the long run. It makes it harder for us to standardize on an IR between circuit and pulse. For example, what happens if a circuit call out to a function or equivalently jumpts to a label in the program. This tuple format doesn't provide a location for that function/label to go.

If alternatively 
```python
class Instruction(abc.ABC)

def __init__(self, operands, metadata)

class CircuitInstruction
     def __init__(self, qubits, clbits):
          super().__init__([qubits, clbits])
```
This gives us a path forward to IR unification.